Update implementation to meet agreed requirements in 2022-06-23 meeting

HassanAkbar opened the pull request:
### Metanorma PR checklist

 - [ ] We only keep ENTITY definitions in ISO 10303-2. Remove the TYPE, FUNCTION, RULE, constant definitions/descriptions from the document. This information is only provided in the EXPRESS code.

 - [x] As every Clause 4/5 term entry has a source, remove the NOTE, and add the “incompletely defined in this document” sentence to Clause 4 and 5 at the top.

 - [x] Alter definition format to use a generated sentence directly from EXPRESS, keep descriptions.xml text (the current text we have) as a Note to entry of the term.

 - [x] Add new Test cases and fix failing ones.

### Example Output

```adoc

== Surface_conditions_arm

This term is incompletely defined in this document.
// STEPmod path: modules/surface_conditions/arm_descriptions.xml

=== sc_additional_application_domain_enumeration

domain:[module: Surface_conditions_arm]

entity data type that represents a/an sc_additional_application_domain_enumeration entity

NOTE: The *sc_additional_application_domain_enumeration* provides predefined identifiers for the additional application domains of *Product_view_definition* s.


[.source]
<<ISO-TS_10303-1110_2021>>


// STEPmod path: modules/surface_conditions/arm_descriptions.xml

=== surface_lay_and_orientation

domain:[module: Surface_conditions_arm]

entity data type that represents a/an surface_lay_and_orientation entity

NOTE: The *surface_lay_and_orientation* provides predefined identifiers of the surface lay as specified in table 2 of ISO 1302:2002.


[.source]
<<ISO-TS_10303-1110_2021>>

...
```
 
closes #123

HassanAkbar commented:
@ronaldtse  I have a question related to this, currently we are iterating on the `ext_description` in `descriptions.xml` file and then finding the related `ENTITY` in the `.exp` file to generate the `description`, but the issue is there are some cases where there is an `ext_description` present but not the coresponding `ENTITY` (e.g `restrict_support_structure_category` has `ext_description` inside `modules/additive_manufacturing_part_and_build_information/mim_descriptions.xml` but there is no `ENTITY Additive_manufacturing_part_and_build_information`), so in that case I am using the `No SUBTYPE` template for definition.

Is this approach correct or now after switching to `.exp` files we should iterate on `ENTITY` and find the coresponding `ext_description` for the the `note`?

TRThurman commented:
As noted there can be extra information in the description xml file.
So we must iterate over the EXPRESS and then find the description.
Once the algorithm is updated, does the follow-on question still exist?

> On Aug 10, 2022, at 8:33 AM, HassanAkbar ***@***.***> wrote:
> 
> 
> @ronaldtse <https://github.com/ronaldtse> I have a question related to this, currently we are iterating on the ext_description in descriptions.xml file and then finding the related ENTITY in the .exp file to generate the description, but the issue is there are some cases where there is an ext_description present but not the coresponding ENTITY (e.g restrict_support_structure_category has ext_description inside modules/additive_manufacturing_part_and_build_information/mim_descriptions.xml but there is no ENTITY Additive_manufacturing_part_and_build_information), so in that case I am using the No SUBTYPE template for definition.
> 
> Is this approach correct or now after switching to .exp files we should iterate on ENTITY and find the coresponding ext_description for the the note?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/metanorma/stepmod-utils/pull/124#issuecomment-1210680619>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAMMKVCQGVJJVAPDRSQR5CTVYOVR7ANCNFSM56ELYY2A>.
> You are receiving this because you are subscribed to this thread.
>

ronaldtse commented on exe/stepmod-extract-terms:
```diff
@@ -136,6 +136,7 @@ part_resources.each do |(bibdata, current_part_resources)|
   fn = "04x-stepmod-entities-resources-#{bibdata.part}.adoc"
   File.open(fn, "w") do |file|
     # file.puts("== #{part_to_title(bibdata)}\n\n")
+    file.puts("This term is incompletely defined in this document.\n")
```
Remove this line -- not necessary.

ronaldtse commented on exe/stepmod-extract-terms:
```diff
@@ -164,6 +165,7 @@ part_modules.each do |(bibdata, part_modules_arm, part_modules_mim)|
 
       # puts "SCHEMA NAME ARM: #{schema_name}"
       file.puts("== #{schema_name}\n\n")
+      file.puts("This term is incompletely defined in this document.\n")
```
Remove, this line will break the document schema.

ronaldtse commented on lib/stepmod/utils/concept.rb:
```diff
@@ -11,25 +13,26 @@ class Concept < Glossarist::LocalizedConcept
           if definition_xml.name == "ext_description"
-            converted_definition = <<~TEXT
-              #{converted_definition}
+            # converted_definition = <<~TEXT
```
The `converted_definition` should become this:

```adoc
NOTE: #{converted_definition}

NOTE: This term is incompletely defined in this document.
Reference <<#{reference_anchor}>> for the complete definition.
```

ronaldtse replied:
Thank you!

TRThurman commented:
clarifications:
1- in the domain where it is a 'module', the correct term is 'application module'.
2- document source is to be converted to a comment and retained in the asciidoc file.
This:
[.source]
<<ISO-TS_10303-1110_2021>>

becomes this:

//[.source]
//<<ISO-TS_10303-1110_2021>>


> On Aug 10, 2022, at 7:37 AM, HassanAkbar ***@***.***> wrote:
> 
> Metanorma PR checklist
> 
>  We only keep ENTITY definitions in ISO 10303-2. Remove the TYPE, FUNCTION, RULE, constant definitions/descriptions from the document. This information is only provided in the EXPRESS code.
> 
>  As every Clause 4/5 term entry has a source, remove the NOTE, and add the “incompletely defined in this document” sentence to Clause 4 and 5 at the top.
> 
>  Alter definition format to use a generated sentence directly from EXPRESS, keep descriptions.xml text (the current text we have) as a Note to entry of the term.
> 
>  Add new Test cases and fix failing ones.
> 
> Example Output
> 
> == Surface_conditions_arm
> 
> This term is incompletely defined in this document.
> // STEPmod path: modules/surface_conditions/arm_descriptions.xml
> 
> === sc_additional_application_domain_enumeration
> 
> domain:[module: Surface_conditions_arm]
> 
> entity data type that represents a/an sc_additional_application_domain_enumeration entity
> 
> NOTE: The *sc_additional_application_domain_enumeration* provides predefined identifiers for the additional application domains of *Product_view_definition* s.
> 
> 
> [.source]
> <<ISO-TS_10303-1110_2021>>
> 
> 
> // STEPmod path: modules/surface_conditions/arm_descriptions.xml
> 
> === surface_lay_and_orientation
> 
> domain:[module: Surface_conditions_arm]
> 
> entity data type that represents a/an surface_lay_and_orientation entity
> 
> NOTE: The *surface_lay_and_orientation* provides predefined identifiers of the surface lay as specified in table 2 of ISO 1302:2002.
> 
> 
> [.source]
> <<ISO-TS_10303-1110_2021>>
> 
> ...
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/metanorma/stepmod-utils/pull/124 <https://github.com/metanorma/stepmod-utils/pull/124>
> Commit Summary
> 
> def7db7 <https://github.com/metanorma/stepmod-utils/pull/124/commits/def7db74c89dc51dc9abe293132a07c428b86ff2> Update implementation to meet agreed requirements in 2022-06-23 meeting
> File Changes (5 files <https://github.com/metanorma/stepmod-utils/pull/124/files>)
> M exe/stepmod-extract-terms <https://github.com/metanorma/stepmod-utils/pull/124/files#diff-1eca87f0f60e873fd7473d31e131c0fe596615361d330fb3c3c80a44127e163e> (2)
> M lib/stepmod/utils/concept.rb <https://github.com/metanorma/stepmod-utils/pull/124/files#diff-0eaa69f0e731fa8393f4ece8ab6f1b7119d0203fcbd929f71523ffd2d02ce434> (15)
> M lib/stepmod/utils/converters/stepmod_ext_description.rb <https://github.com/metanorma/stepmod-utils/pull/124/files#diff-5f0697be37f4b3a20634cc381cb20a70515b3ff3ed402d8f74665c4a7e3f0d82> (33)
> M lib/stepmod/utils/terms_extractor.rb <https://github.com/metanorma/stepmod-utils/pull/124/files#diff-0f4e9c7bed6b30b4c150982dfe6548f65616415632b11797fc1ff3f6e6fd4508> (13)
> M stepmod-utils.gemspec <https://github.com/metanorma/stepmod-utils/pull/124/files#diff-0137bc0380cd5e5b8b96dadf88158b7419e2046eea7023a14847ce34d77fcbf3> (1)
> Patch Links:
> 
> https://github.com/metanorma/stepmod-utils/pull/124.patch <https://github.com/metanorma/stepmod-utils/pull/124.patch>
> https://github.com/metanorma/stepmod-utils/pull/124.diff <https://github.com/metanorma/stepmod-utils/pull/124.diff>
> —
> Reply to this email directly, view it on GitHub <https://github.com/metanorma/stepmod-utils/pull/124>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAMMKVGFSYYYNSXNWYNV743VYOO75ANCNFSM56ELYY2A>.
> You are receiving this because you are subscribed to this thread.
>

ronaldtse commented:
@HassanAkbar as mentioned by @TRThurman, let's reverse the iteration so that:
1. Iterate over the EXPRESS schema (`.exp`), for all ENTITYs
2. For every ENTITY, look up the corresponding `<ext_description>` in the `descriptions.xml` file
3. Render the term/definition.

TRThurman commented:
@ronaldtse, thanks! My comment on commenting out the [.source] is incorrect for clauses 4, 5.

Sorry for confusion. 

> On Aug 10, 2022, at 8:50 AM, Ronald Tse ***@***.***> wrote:
> 
> 
> @ronaldtse requested changes on this pull request.
> 
> In exe/stepmod-extract-terms <https://github.com/metanorma/stepmod-utils/pull/124#discussion_r942469587>:
> 
> > @@ -136,6 +136,7 @@ part_resources.each do |(bibdata, current_part_resources)|
>    fn = "04x-stepmod-entities-resources-#{bibdata.part}.adoc"
>    File.open(fn, "w") do |file|
>      # file.puts("== #{part_to_title(bibdata)}\n\n")
> +    file.puts("This term is incompletely defined in this document.\n")
> Remove this line -- not necessary.
> 
> In exe/stepmod-extract-terms <https://github.com/metanorma/stepmod-utils/pull/124#discussion_r942470295>:
> 
> > @@ -164,6 +165,7 @@ part_modules.each do |(bibdata, part_modules_arm, part_modules_mim)|
>  
>        # puts "SCHEMA NAME ARM: #{schema_name}"
>        file.puts("== #{schema_name}\n\n")
> +      file.puts("This term is incompletely defined in this document.\n")
> Remove, this line will break the document schema.
> 
> In lib/stepmod/utils/concept.rb <https://github.com/metanorma/stepmod-utils/pull/124#discussion_r942471588>:
> 
> >              },
>            )
>  
>            return nil if converted_definition.nil? || converted_definition.strip.empty?
>  
>            if definition_xml.name == "ext_description"
> -            converted_definition = <<~TEXT
> -              #{converted_definition}
> +            # converted_definition = <<~TEXT
> The converted_definition should become this:
> 
> NOTE: #{converted_definition}
> 
> NOTE: This term is incompletely defined in this document.
> Reference <<#{reference_anchor}>> for the complete definition.
> —
> Reply to this email directly, view it on GitHub <https://github.com/metanorma/stepmod-utils/pull/124#pullrequestreview-1068307957>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAMMKVAAB2ETPVYJMIHVL5TVYOXR3ANCNFSM56ELYY2A>.
> You are receiving this because you were mentioned.
>

ronaldtse commented:
@TRThurman but your message did remind me about commenting out for Clause 3 -- I haven't actually updated stepmod-utils to do this but have done the commenting out manually. Now that we have finalized that Clause 3 terms belongs to the document, we will do the commenting out.

ronaldtse commented on lib/stepmod/utils/converters/stepmod_ext_description.rb:
```diff
@@ -43,21 +43,46 @@ def convert(node, state = {})
+          if entity.nil?
+            "entity data type that represents a/an #{entity_name} entity"
+          elsif entity.subtype_of.size.zero?
+            "entity data type that represents a/an #{entity.id} entity"
```
We will have to select "an" or "a" using some detection mechanism.

ronaldtse commented on lib/stepmod/utils/converters/stepmod_ext_description.rb:
```diff
@@ -43,21 +43,46 @@ def convert(node, state = {})
+          elsif entity.subtype_of.size == 1
+            "entity data type that is a type of #{entity.subtype_of[0].id} that represents a/an #{entity.id} entity"
+          else
+            "entity data type that is a type of #{entity.subtype_of.map(&:id).join(' and ')} that represents a/an #{entity.id} entity"
```
As above re: a, an.

ronaldtse commented on lib/stepmod/utils/converters/stepmod_ext_description.rb:
```diff
@@ -43,21 +43,46 @@ def convert(node, state = {})
+          elsif entity.subtype_of.size.zero?
+            "entity data type that represents a/an #{entity.id} entity"
+          elsif entity.subtype_of.size == 1
+            "entity data type that is a type of #{entity.subtype_of[0].id} that represents a/an #{entity.id} entity"
```
As above re: a, an.

ronaldtse commented on lib/stepmod/utils/converters/stepmod_ext_description.rb:
```diff
@@ -43,21 +43,46 @@ def convert(node, state = {})
+
+        def extract_entity_text(entity, entity_name)
+          if entity.nil?
+            "entity data type that represents a/an #{entity_name} entity"
```
Need to address "a", "an".

ronaldtse commented on spec/stepmod/concept_spec.rb:
```diff
@@ -85,9 +87,11 @@
+        entity data type that represents a/an complement entity
 
-        A *complement* is a relationship that is between
+        NOTE: A *complement* is a relationship that is between
```
This NOTE is a long one that contains blank lines. Blank lines causes Metanorma to split text into multiple paragraphs. In this case, we need to use an alternative syntax of NOTE.

The following three forms are equivalent:

Single paragraph:
```adoc
NOTE: This is a note.
```

Continuation style:
```adoc
NOTE: This is a note.
+
* list in a note.
```

Block style:
```adoc
[NOTE]
--
This is a note.

* list in a note.
--
```

We need to use the block-style syntax (the third one).

ronaldtse commented on spec/stepmod/terms_extractor_spec.rb:
```diff
@@ -40,39 +40,33 @@
 
-        NOTE: This term is incompletely defined in this document.
-        Reference <<ISO-TS_10303-1047_2014>> for the complete definition.
+        NOTE: An *applied_action_assignment* is an <<express:action_schema.action,action>> related to the data that are affected by the <<express:action_schema.action,action>>. An *applied_action_assignment* is a type of <<express:management_resources_schema.action_assignment,action_assignment>>.
```
We need to add one additional NOTE (a second NOTE) to say that:

```adoc
NOTE: This term is incompletely defined in this document. Reference <<ISO-TS_10303-1047_2014>> for the complete definition.
```

ronaldtse commented on spec/stepmod/terms_extractor_spec.rb:
```diff
@@ -40,39 +40,33 @@
 
-        NOTE: This term is incompletely defined in this document.
-        Reference <<ISO-TS_10303-1047_2014>> for the complete definition.
+        NOTE: An *Activity* is the identification of the occurrence of an action that has taken place, is taking place, or is expected to take place in the future. The procedure executed during that *Activity* is identified with the <<express:Activity_method_arm.Activity_method,Activity_method>> that is referred to by the *chosen_method* attribute.
```
Need to add an additional NOTE below this one:
```adoc
NOTE: This term is incompletely defined in this document. Reference <<ISO-TS_10303-1047_2014>> for the complete definition.
```

ronaldtse commented on spec/stepmod/terms_extractor_spec.rb:
```diff
@@ -40,39 +40,33 @@
-
-        NOTE: This term is incompletely defined in this document.
-        Reference <<ISO_10303-41_2021>> for the complete definition.
+        NOTE: An *action_directive* is an authoritative instrument that provides directions to achieve the specified results.
```
See below, need to add the 'incompletely defined' NOTE as a second NOTE.

ronaldtse commented on spec/stepmod/concept_spec.rb:
```diff
@@ -96,10 +100,6 @@
         that indicates set S2 consists of all members of U that are not members of S1.
 
 
-        NOTE: This term is incompletely defined in this document.
```
We still need to keep this NOTE for items in Clause 4 (`04x-*`) and Clause 5 (`05x-*.adoc`).

HassanAkbar replied:
> 3. As every Clause 4/5 term entry has a source, remove the NOTE, and add the “incompletely defined in this document” sentence to Clause 4 and 5 at the top.

@ronaldtse this is mentioned in the [ticket](https://github.com/metanorma/stepmod-utils/issues/123), are we not doing this right now or am I missing something here?

ronaldtse replied:
Yes you are correct. I got confused. Instead please ensure we have the “[.source]” entry for the Clause 4 and Clause 5. Thanks.

HassanAkbar commented:
@ronaldtse there are some entities in `.exp` files that do not have a corresponding `ext_description` in `.xml` file, e.g [`modules/cardinality_of_relationship`](https://github.com/metanorma/iso-10303-stepmod-wg12/tree/de05521fedcadb45f7d86431bb614eb6b8b71d60/data/modules/cardinality_of_relationship) does not have a `arm_descriptions.xml` file which contains `ext_descriptions` for `Cardinality_of_relationship_arm` schema.

What should be the note in those cases?

TRThurman commented:
This is a case where the descriptions are in arm.xml.
There are also descriptions in mim.xml


> On Aug 16, 2022, at 5:56 AM, HassanAkbar ***@***.***> wrote:
> 
> 
> @ronaldtse <https://github.com/ronaldtse> there are some entities in .exp files that do not have a corresponding ext_description in .xml file, e.g modules/cardinality_of_relationship <https://github.com/metanorma/iso-10303-stepmod-wg12/tree/de05521fedcadb45f7d86431bb614eb6b8b71d60/data/modules/cardinality_of_relationship> does not have a arm_descriptions.xml file which contains ext_descriptions for Cardinality_of_relationship_arm schema.
> 
> What should be the note in those cases?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/metanorma/stepmod-utils/pull/124#issuecomment-1216478119>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAMMKVH2A46DIUI7WV36LXLVZNXXRANCNFSM56ELYY2A>.
> You are receiving this because you were mentioned.
>

HassanAkbar commented:
In the existing code we were only looking in `arm_descriptions.xml` and `mim_desctiptions.xml` both of which are not present in case of [`modules/cardinality_of_relationship`](https://github.com/metanorma/iso-10303-stepmod-wg12/tree/de05521fedcadb45f7d86431bb614eb6b8b71d60/data/modules/cardinality_of_relationship). However `arm.xml` and `mim.xml` files are present. If we want to use these then the flow will be something like below

**For `module` folder**
1. Loop on entities in `arm.exp` file
2. Find the corresponding entity in `arm_description.xml` file if it exists.
3. If `arm_description.xml` file does not exist then look for definition in `arm.xml` file
4. Loop through entities in `mim.exp` file that are not found in the previous loop
5. Find the corresponding entity in `mim_description.xml` file if it exists.
6. If `mim_description.xml` file does not exist then look for definition in `mim.xml` file

Is this flow correct?

TRThurman commented:
Yes

> On Aug 16, 2022, at 8:19 AM, HassanAkbar ***@***.***> wrote:
> 
> 
> In the existing code we were only looking in arm_descriptions.xml and mim_desctiptions.xml both of which are not present in case of modules/cardinality_of_relationship <https://github.com/metanorma/iso-10303-stepmod-wg12/tree/de05521fedcadb45f7d86431bb614eb6b8b71d60/data/modules/cardinality_of_relationship>. However arm.xml and mim.xml files are present. If we want to use these then the flow will be something like below
> 
> For module folder
> 
> Loop on entities in arm.exp file
> Find the corresponding entity in arm_description.xml file if it exists.
> If arm_description.xml file does not exist then look for definition in arm.xml file
> Loop through entities in mim.exp file that are not found in the previous loop
> Find the corresponding entity in mim_description.xml file if it exists.
> If mim_description.xml file does not exist then look for definition in mim.xml file
> Is this flow correct?
> 
> —
> Reply to this email directly, view it on GitHub <https://github.com/metanorma/stepmod-utils/pull/124#issuecomment-1216628640>, or unsubscribe <https://github.com/notifications/unsubscribe-auth/AAMMKVEI3FL6DMG64Y6VINLVZOIORANCNFSM56ELYY2A>.
> You are receiving this because you were mentioned.
>

ronaldtse commented:
@HassanAkbar actually there is an issue when we work with the XML files. Somehow when I issued the instructions, I forgot about https://github.com/metanorma/iso-10303-2/issues/94.

The goal here is actually **NOT** to work with XML files. The goal here is to directly read in [Annotated EXPRESS](https://github.com/metanorma/annotated-express/) files.

For "Resources":
* e.g. for `resources/action_schema`, 
    * Original: https://github.com/metanorma/iso-10303-stepmod/blob/main/data/resources/action_schema/action_schema.exp
    * Annotated: https://github.com/metanorma/iso-10303-stepmod/blob/main/data/resources/action_schema/action_schema_annotated.exp
* The Annotated EXPRESS files already incorporate all the content from `descriptions.xml`
* This means that by using Expressir you can already obtain all the information.

For "Modules":
* e.g. for `modules/cardinality_of_relationship`,
    * Original:
      * https://github.com/metanorma/iso-10303-stepmod/blob/main/data/modules/cardinality_of_relationship/arm.exp
      * https://github.com/metanorma/iso-10303-stepmod/blob/main/data/modules/cardinality_of_relationship/mim.exp
    * Annotated:
      * MISSING! ( to be implemented here: https://github.com/metanorma/stepmod-utils/blob/main/exe/stepmod-annotate-all)
* The Annotated EXPRESS files *should* already incorporate all the content from `arm.xml` and `mim.xml`.

Here's how the flow works:
1. We pull the STEPmod repository (iso-10303-stepmod-wg12) using this workflow https://github.com/metanorma/iso-10303-stepmod-migration/blob/main/.github/workflows/annotate.yml
2. The workflow executes stepmod-utils (this gem) and runs `stepmod-annotate-all` to merge all *.xml content into Annotated EXPRESS files (.exp files but with structured comments)
3. The workflow pushes the repository with the added annotations back

So the task here @HassanAkbar is actually:
1. For Resources: change the code here to read from the resources EXPRESS files
2. For Modules: update `stepmod-annotate-all` to also cover the conversion of Modules to Annotated EXPRESS
3. For Modules: change the code here to read from the modules EXPRESS files

HassanAkbar commented:
@ronaldtse The definition for the activity in the annotated file is
```
An *Activity* is the identification of the occurrence of an action that has taken place, is taking place, or is expected to take place in the future. The procedure executed during that *Activity* is identified with the <<express:Activity_method_arm.Activity_method,Activity_method>> that is referred to by the *chosen_method* attribute.
[example]
====
Change, distilling, design, a process to drill a hole, and a task such as training someone, are examples of activities.
====

[NOTE]
--
Status information identifying the level of completion of each activity may be provided within an instance of <<express:Activity_arm.Activity_status,Activity_status>>.
--

[NOTE]
--
The items that are affected by an *Activity*, for example as input or output, may be identified within an instance of <<express:Activity_arm.Applied_activity_assignment,Applied_activity_assignment>>.
--
```

As you can see it includes `notes` and `examples`. When I add this in the new definition as a `note` the existing `notes` the formatting is distorted. How do you suggest we resolve this?

ronaldtse commented:
You need a line break after the definition paragraph and `[example]`.

HassanAkbar commented:
@ronaldtse this is the old definition when I add it as a note to the new definition like this

```
```

ronaldtse commented:
We have to fix the format. The correct representation is below:

```adoc
=== Activity
domain:[application module: Activity_arm]

entity data type that represents an Activity entity

[example]
====
Change, distilling, design, a process to drill a hole, and a task such as training someone, are examples of activities.
====

[NOTE]
--
An *Activity* is the identification of the occurrence of an action that has taken place, is taking place, or is expected to take place in the future. The procedure executed during that *Activity* is identified with the <<express:Activity_method_arm.Activity_method,Activity_method>> that is referred to by the *chosen_method* attribute.
--

[NOTE]
--
Status information identifying the level of completion of each activity may be provided within an instance of <<express:Activity_arm.Activity_status,Activity_status>>.
--

[NOTE]
--
The items that are affected by an *Activity*, for example as input or output, may be identified within an instance of <<express:Activity_arm.Applied_activity_assignment,Applied_activity_assignment>>.
--
```

HassanAkbar commented:
@ronaldtse I think I am not able to explain the problem clearly. The flow that we need to follow is

- Read from annotated EXPRESS file using `expressir` and create a definition.

The annotated EXPRESS file for `activity`(arm_annotated.exp) contains 

```
(*
$Id: arm.exp,v 1.18 2009/12/11 13:39:31 mikeward Exp $
ISO TC184/SC4/WG12 N2941 - ISO/TS 10303-1047 Activity - EXPRESS ARM
*) 


SCHEMA Activity_arm;

USE FROM Activity_method_arm;    -- ISO/TS 10303-1049


TYPE activity_item = EXTENSIBLE GENERIC_ENTITY SELECT;
END_TYPE; 

ENTITY Activity;
  id : STRING;
  name : STRING;
  description : OPTIONAL STRING;
  chosen_method : Activity_method;
END_ENTITY;

...

END_SCHEMA;  -- Activity_arm


(*"Activity_arm.Activity"
An *Activity* is the identification of the occurrence of an action that has taken place, is taking place, or is expected to take place in the future. The procedure executed during that *Activity* is identified with the <<express:Activity_method_arm.Activity_method,Activity_method>> that is referred to by the *chosen_method* attribute. 
[example]
====
Change, distilling, design, a process to drill a hole, and a task such as training someone, are examples of activities.
====

[NOTE]
--
Status information identifying the level of completion of each activity may be provided within an instance of <<express:Activity_arm.Activity_status,Activity_status>>.
--

[NOTE]
--
The items that are affected by an *Activity*, for example as input or output, may be identified within an instance of <<express:Activity_arm.Applied_activity_assignment,Applied_activity_assignment>>.
--
*)

(*"Activity_arm.Activity.id"
the identifier that distinguishes the *Activity*.
*)

(*"Activity_arm.Activity.name"
the words by which the *Activity* is known.
*)

...

```

The issue is when I read the old definition it contains `examples` and `notes`. 
Removing the `notes` and `examples` from the `old definition` and adding them to the new definition will require some string manipulations or again reading the `xml` files because currently, I didn't find a way to extract these from the express file.

ronaldtse commented:
In today's ISO/TC 184/SC 4/TF 1 meeting Keith agreed with the 2nd approach, which is to treat NOTEs, EXAMPLEs etc as separate Annotated EXPRESS tagged remarks. We will need to update the annotation generator.

HassanAkbar commented:
@ronaldtse I've updated the annotation generator and now the annotated express file contains the following

```
(*"Activity_arm.Activity"
An **Activity** is the identification of the occurrence of an action that has taken place, is taking place, or is expected to take place in the future. The procedure executed during that **Activity** is identified with the <<express:Activity_method_arm.Activity_method,Activity_method>> that is referred to by the **chosen_method** attribute.
*)



(*"Activity_arm.Activity.__example"
Change, distilling, design, a process to drill a hole, and a task such as training someone, are examples of activities.
*)


(*"Activity_arm.Activity.__note"
Status information identifying the level of completion of each activity may be provided within an instance of <<express:Activity_arm.Activity_status,Activity_status>>.
*)



(*"Activity_arm.Activity.__note"
The items that are affected by an *Activity*, for example as input or output, may be identified within an instance of <<express:Activity_arm.Applied_activity_assignment,Applied_activity_assignment>>.
*)
```

My question is do we need to include the `notes` and `examples` as `notes` and `examples` in the new definition?

ronaldtse commented:
Yes, we need to translate those into:
```adoc
[NOTE]
--
...
--
```

and 
```adoc
[example]
====
...
====
```

HassanAkbar commented:
Achieved in this PR

- As every Clause 4/5 term entry has a source, Removed the NOTE “incompletely defined in this document”.
- Updated `stepmod-annotate-all` to add note as `<entity>.__note` and example as `<entity>.__example` in annotated express files.
- Updated extract terms flow to read all the old definitions, notes, and examples from the `annotated_express` files and 
- Generated definition from entity name and template based on subtypes.
- Addd old definition as a `note` to new definition.
- Updated test cases accordingly.

ronaldtse commented on spec/stepmod/terms_extractor_spec.rb:
```diff
@@ -39,46 +39,66 @@
-
-        NOTE: This term is incompletely defined in this document.
-        Reference <<ISO_10303-41_2021>> for the complete definition.
+        [EXAMPLE]
```
We want lowercase `[example]` instead of `[EXAMPLE]`.

ronaldtse replied:
Fixed.

ronaldtse reviewed (changes requested):
Looks good! Only one nit.

ronaldtse commented:
Once this is merged we should run it on https://github.com/metanorma/iso-10303-2

ronaldtse commented:
We need a release bump @HassanAkbar .

ronaldtse commented:
Actually this implementation is incomplete:
```
(*"action_schema.supported_item"
The **supported_item** allows for the designation of an <<express:action_schema.action_directive,action_directive>>, an <<express:action_schema.action,action>>, or an <<express:action_schema.action_method,action_method>>.
*)

(*".__note"
This specifies the use of an <<express:action_schema.action_resource,action_resource>>.
*)
```

Notice that the name of the `__note` is missing.
